src/Control/Carrier/Cut/Church.hs

```diff
@@ -92,5 +91,5 @@ instance (Algebra sig m, Effect sig) => Algebra (Cut :+: NonDet :+: sig) (CutC m
   eff (L (Call m k)) = CutC $ \ cons nil fail -> runCut (\ a as -> runCut cons as fail (k a)) nil nil m
   eff (R (L (L Empty)))      = empty
   eff (R (L (R (Choose k)))) = k True <|> k False
-  eff (R (R other))          = CutC $ \ cons nil _ -> eff (handle [()] (fmap concat . traverse runCutA) other) >>= foldr cons nil
+  eff (R (R other))          = CutC $ \ cons nil fail -> eff (handle (pure ()) (runCut (fmap . (<|>)) (pure empty) (pure cutfail)) other) >>= runCut cons nil fail
```

Comment:
This trick worked out beautifully, and I intend to steal it for the other nondeterminism carriers in #286 (suitably repurposed): basically, we’re using `CutC m` as its _own_ state functor. If you’ll note, we use a very similar trick in the `MonadFix` instance, using `CutC PureC` as a representation of the fixpoint.

In both cases, this works precisely because it allows us to encode lists with soft and hard emptiness: i.e. lists differentiating between `empty` and `cutfail`. This means that we then propagate `cutfail`s through the fixpoint and higher-order actions correctly, instead of mapping them to `empty` and losing the essential information.